_posts/2022-05-18-torchopt/torchopt.Rmd

```diff
@@ -0,0 +1,356 @@
+
+[^4]: Alluding to strange attractors here. If you're interested, you might want to check out the mini-series on chaos and deep learning: [Deep attractors: Where deep learning meets chaos](https://blogs.rstudio.com/ai/posts/2020-06-24-deep-attractors/), [Time series prediction with FNN-LSTM](https://blogs.rstudio.com/ai/posts/2020-07-20-fnn-lstm/), and [FNN-VAE for noisy time series forecasting](https://blogs.rstudio.com/ai/posts/2020-07-31-fnn-vae-for-noisy-timeseries/).
+
+## A second-order optimizer for neural networks: ADAHESSIAN
+
+On to the one algorithm I'd like to check out specifically. Subsequent to just a little bit of learning-rate experimentation, I was able to arrive at an excellent result after just thirty-five steps.
+
+```{r}
+test_optim(
+    optim = optim_adahessian,
+    opt_hparams = list(lr = 0.3),
+    test_fn = list(flower, function() (c(x0 = 20, y0 = 20, xmax = 3, xmin = -3, ymax = 3, ymin = -3))),
+    steps = 35
+)
+```
+
+![Minimizing the flower function with ADAHESSIAN. Setup no. 1: lr = 0.3, 35 steps.](images/adahessian-steps-35.png){alt="Minimizing the flower function with AdamW. Setup no. 3: lr"}
+
+Given our recent experiences with AdamW though -- meaning, its "just not settling in" very close to the minimum -- we may want to run an equivalent test with ADAHESSIAN, as well. What happens if we go on optimizing quite a bit longer -- for two-hundred steps, say?
+
+```{r}
+test_optim(
+    optim = optim_adahessian,
+    opt_hparams = list(lr = 0.3),
+    test_fn = list(flower, function() (c(x0 = 20, y0 = 20, xmax = 3, xmin = -3, ymax = 3, ymin = -3))),
+    steps = 200
+)
+```
+
+![Minimizing the flower function with ADAHESSIAN. Setup no. 2: lr = 0.3, 200 steps.](images/adahessian-steps-200.png)
+
+Like AdamW, ADAHESSIAN goes on to "explore" the petals, but it does not stray as far away from the minimum.
+
+Is this surprising? I wouldn't say it is. The argument is the same as with AdamW, above: Its algorithm has been tuned to perform well on large neural networks, not to solve a classic, hand-crafted minimization task.
+
+Now we've heard that argument twice already, it's time to verify the explicit assumption: that a classic second-order algorithm handles this better. In other words, it's time to revisit L-BFGS.
+
+## Best of the classics: Revisiting L-BFGS
+
+To use `test_optim()` with L-BFGS, we need to take a little detour. If you've read the post on L-BFGS, you may remember that with this optimizer, it is necessary to wrap both the call to the test function and the evaluation of the gradient in a closure. (The reason being that both have to be callable several times per iteration.)
+
+Now, seeing how L-BFGS is a very special case, and few people are likely to use `test_optim()` with it in the future, it wouldn't seem worthwhile to make that function handle different cases. For this on-off test, I simply copied and modified the code as required. The result, `test_optim_lbfgs()`, is found in the appendix.
```

Comment:
I believe you can link to the appendix (it's an anchored header). So, I'd  add that where you mention the appendix.